src/midi/midi_mt32_model.cpp

```diff
@@ -0,0 +1,145 @@
+/*
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *
+ *  Copyright (C) 2021-2021  The DOSBox Staging Team
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along
+ *  with this program; if not, write to the Free Software Foundation, Inc.,
+ *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+#include "midi_mt32_model.h"
+
+#if C_MT32EMU
+
+#include <cassert>
+
+#include "fs_utils.h"
+
+// Construct a new model and ensure both PCM and control ROM(s) are provided
+Model::Model(const std::string &rom_name,
+             const Rom *pcm_full,
+             const Rom *pcm_a,
+             const Rom *pcm_b,
+             const Rom *ctrl_full,
+             const Rom *ctrl_a,
+             const Rom *ctrl_b)
+        : name(rom_name),
+          pcmFull(pcm_full),
+          pcmA(pcm_a),
+          pcmB(pcm_b),
+          ctrlFull(ctrl_full),
+          ctrlA(ctrl_a),
+          ctrlB(ctrl_b)
+{
+	assert(!name.empty());
+	assert(pcmFull || (pcmA && pcmB));
+	assert(ctrlFull || (ctrlA && ctrlB));
+}
+
+// Checks if its ROMs can be positively found in the provided directory
+bool Model::InDir(const service_t &service, const std::string &dir)
+{
+	assert(service);
+	if (inDir.find(dir) != inDir.end())
+		return inDir[dir];
+
+	auto find_rom = [&service, &dir](const Rom *Rom) -> bool {
+		if (!Rom)
+			return false;
+
+		const std::string rom_path = dir + Rom->filename;
+		if (!path_exists(rom_path))
+			return false;
+
+		mt32emu_rom_info info;
+		if (service->identifyROMFile(&info, rom_path.c_str(), nullptr) != MT32EMU_RC_OK)
+			return false;
+
+		if (Rom->type == ROM_TYPE::UNVERSIONED)
+			return true;
+
+		const bool found_pcm = info.pcm_rom_id && (Rom->id == info.pcm_rom_id);
+		const bool found_ctrl = info.control_rom_id && (Rom->id == info.control_rom_id);
+		return found_pcm || found_ctrl;
+	};
+
+	auto find_both = [find_rom](const Rom *rom_a, const Rom *rom_b) -> bool {
+		return find_rom(rom_a) && find_rom(rom_b);
+	};
+
+	const bool have_pcm = find_rom(pcmFull) || find_both(pcmA, pcmB);
+	const bool have_ctrl = find_rom(ctrlFull) || find_both(ctrlA, ctrlB);
```

Comment:
I find this implementation very nice and readable :)

Perhaps we can drop `find_both` and simply write:

```C++
const bool have_pcm = find_rom(pcmFull) || (find_rom(pcmA) && find_rom(pcmB));
```
but it's nitpicking - code is fine already.

I would prefer if we didn't mix `camelCase` in field names with `snake_case` in local variables - can we change field names to use snake_case as well?